### ISSUE TITLE
JS development cycle is broken after dropwizard merge

### ISSUE THREAD

**Issue by karussell** [graphhopper/graphhopper#1330]
In order to develop JS code we did:

```
cd web
npm run watch
```

and then we were able to refresh the browser. Now the changes take effect only if we repackage the jar e.g. via

```
mvn clean install -DskipTests=true
```

It is possible to load assets directly from disc instead of looking inside the jar. I found [this plugin](https://github.com/dirkraft/dropwizard-file-assets#integration)

With a changed npm-`watch` task one can use it if the server was started from an IDE:
`watchify src/main/resources/assets/js/main-template.js -o target/classes/assets/js/main.js --debug --verbose`
but this is ugly as we don't want to force JS devs to mess around with IDEs.

**Comment by karussell** [comment]
Using the linked plugin and

```java
bootstrap.addBundle(new FileAssetsBundle("web/src/main/resources/assets", "/maps/", "index.html"));
```

works as intended.

@michaz can we make this line configurable somehow and avoid as default (for production) to increase security?

**Comment by karussell** [comment]
I was able to make it working through a new application class:
```
public class GraphHopperDebugApplication extends GraphHopperApplication {
    public static void main(String[] args) throws Exception {
        new GraphHopperDebugApplication().run(args);
    }

    @Override
    public void initialize(Bootstrap<GraphHopperServerConfiguration> bootstrap) {
        super.initialize(bootstrap);
        bootstrap.addBundle(new FileAssetsBundle("web/src/main/resources/assets", "/maps/", "index.html"));
    }
}
```

And called `java -cp gh.jar com.graphhopper.http.GraphHopperDebugApplication server config.yml`
But is this the way to go? I would have thought I could just use a different command like `server-debug` but wasn't able to get it working.

**Comment by ammagamma** [comment]
I tried this but no matter what I did the assets were always picked up from the jar/classpath (although I verified that GraphHopperDebugApplication was running). I would have liked to check the FileAssetServlet debug logging as mentioned [here](https://github.com/dirkraft/dropwizard-file-assets), but then I did not manage to configure log4j with dropwizard :( ? It does not let me pass -Dlog4j.configurationFile via the command line and it seemingly did not pick up any log4j.xml files from the class path (I tried putting it in core/src/main/resources), either. I also did not find out where dropwizard picks up the logging configuration from (?).

**Comment by michaz** [comment]
Log4j is gone -- Dropwizard uses logback. It's configured in-line in the ``config.yml``, see http://www.dropwizard.io/1.3.1/docs/manual/core.html#log-format

**Comment by karussell** [comment]
@ammagamma can you try `./graphhopper.sh webdebug file.pbf`?
Of course you still need to run `cd web; npm run watch` like before.

**Comment by ammagamma** [comment]
I am running `./graphhopper.sh webdebug core/files/monaco.osm`, no changes in config.yml.
Then I create a file `xyz.png` in `web/src/main/resources/assets/` (after running `mvn package` otherwise it ends up in the jar). When I navigate to `localhost:8989/maps/xyz.png` I am getting 404. Before that I tried running `npm run watch` etc. but that should do the same: change some files in the assets folder.

**Comment by karussell** [comment]
Ah, the problem is that there is no automatic copying from assets into targets like its done from npm for the JS file. So that's a bug.

**Comment by karussell** [comment]
No. wait. It should be possible. Will have a look what's wrong. Because the JS file is copied into the same assets folder like before (an intermediate change was to copy it into target/resources)

**Comment by ammagamma** [comment]
I thought the assets folder is simply configured to hold static resources so whatever is in it should be available.

**Comment by karussell** [comment]
(Yes, that's how it should work. But had a couple of other possibililties before so I mixed them up. Sorry. Also it looks like I copied the images into the wrong folder.)

**Comment by karussell** [comment]
Not even
`cp src/main/resources/assets/keep_right.png target/classes/assets/img/`
works :(

**Comment by ammagamma** [comment]
> Log4j is gone

Maybe we could restate the default logging configuration in config.yml (and an entries under 'loggers' for com.graphhopper) to make this clear ? Something like this : 
![image](https://user-images.githubusercontent.com/17603532/38706145-746cb0e4-3eac-11e8-979a-20446a310dcc.png)

**Comment by ammagamma** [comment]
@karussell not sure if I follow the path in GraphHopperDebugApplication is web/src/main/resources/assets not target/classes/assets ?

**Comment by karussell** [comment]
Please ignore me. Have to try this tomorrow ;)

**Comment by karussell** [comment]
Now I even can't get the hot-reloading of JS working. Something is wrong here.

**Comment by karussell** [comment]
Maybe it would be cleaner if we serve the web UI files using a node server like `node http.js` in package.json or via http-server (and `npm install http-server -g`)? We could probably combine watchify and this to avoid three components to start.

**Comment by boldtrn** [comment]
> Maybe it would be cleaner if we serve the web UI files using a node server like 

Yes I agree, I think we could even think about completely separating the web-demo part from this repo. This would IMHO make it easier for web developers to work with the demo. We could try to set this up using the standard node style so that developers have an easy time to find what they are looking for and don't be scared off by this huge java project?

**Comment by karussell** [comment]
I would not split repos as we would make it harder to try out the routing engine, but we could think about moving this into its dedicated `web-ui` or `ui` folder.

**Comment by boldtrn** [comment]
Ok, makes sense as well :+1: .

**Comment by karussell** [comment]
Oh, this was a stupid mistake and is now fixed: https://github.com/graphhopper/graphhopper/commit/4d95eb5b5c7e9adae37e56f3a9c2261e88f65946

@ammagamma would you try again if hot reloading works for you?
@michaz I'll format all files according to new IntelliJ (see linked commit and also stumbled over it [here](https://github.com/graphhopper/graphhopper/pull/1341#discussion_r182512578)) or will this break some of your branches?

**Comment by michaz** [comment]
@karussell Sure, go ahead!

**Comment by michaz** [comment]
> Oh, this was a stupid mistake and is now fixed

Now we are not including the webapp at all anymore in the non-debug-version.

**Comment by karussell** [comment]
> Now we are not including the webapp at all anymore in the non-debug-version.

Bad. Maybe we should then use the configurable asset and include UI by default (to serve via classpath or disc) and offer a separate `api` action without this for stricter security.

**Comment by michaz** [comment]
Here we go. See ``config-example.yml``. No separate class or command necessary. I don't have security concerns. I think this is done.

**Comment by karussell** [comment]
Thanks!

**Comment by karussell** [comment]
Can confirm that this works - have changed the script and docs accordingly.

**Comment by ammagamma** [comment]
> @ammagamma would you try again if hot reloading works for you?

to be honest it does not, but I will try again...

**Comment by karussell** [comment]
Tested it again and when editing the JS file in IntelliJ I didn't even need to save and just refreshed in the browser. Steps I did:

 * latest commit on master
 * ./graphhopper.sh clean; ./graphhopper.sh web /home/peter/Downloads/berlin-latest.osm.pbf
 * (in a second terminal do:) cd web; npm run watch

**Comment by ammagamma** [comment]
Ok it works :) I think the overrides path in my config.yml was not correct relative to the working directory.

**Comment by karussell** [comment]
Ah, ok. Maybe we should change it relative to the root.
